src/masternodes/mn_checks.cpp

```diff
@@ -2992,6 +3008,38 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor
                     if (collateralsLoans.val->ratio() < scheme->ratio)
                         return Res::Err("Vault does not have enough collateralization ratio defined by loan scheme - %d < %d", collateralsLoans.val->ratio(), scheme->ratio);
                 }
+            }
+            if (height >= static_cast<uint32_t>(consensus.GreatWorldHeight)) {
+                const auto loanTokens = mnview.GetLoanTokens(obj.vaultId);
+                if (loanTokens) {
+                    for (const auto& [tokenId, tokenAmount] : loanTokens->balances) {
+                        const auto rate = mnview.GetInterestRate(obj.vaultId, tokenId, height);
+                        assert(rate);
+                        const auto scheme = mnview.GetLoanScheme(vault->schemeId);
+                        assert(scheme);
+                        const auto loanToken = mnview.GetLoanTokenByID(tokenId);
+                        assert(loanToken);
+                        if (const auto totalInterest = TotalInterest(*rate, height, tokenAmount, loanToken->interest, scheme->rate); totalInterest < 0) {
+                            const auto subAmount = tokenAmount > std::abs(totalInterest) ? std::abs(totalInterest) : tokenAmount;
+                            res = mnview.SubLoanToken(obj.vaultId, CTokenAmount{tokenId, subAmount});
+                            if (!res) {
+                                return res;
+                            }
+
+                            res = mnview.SubMintedTokens(tokenId, subAmount);
```

Comment:
those tokens are still in circulation, so they must not be substracted from the minted amount. we only reduced the loan amount. reducing the minted tokens here would add "algorithmic" DUSD that are not accounted for anywhere.

Reply:
Resolved.